File: tests/moonep_planner/reference/planner_reference.cpp
#include "planner_reference.h"

#include <algorithm>
#include <limits>

namespace TileXRMoonEpTest {
namespace {

int64_t AllocIndex(int64_t expert, int32_t dest, int32_t rankSize)
{
    return expert * rankSize + dest;
}

uint32_t Hash(uint32_t value)
{
    value ^= value >> 16;
    value *= 0x7feb352dU;
    value ^= value >> 15;
    value *= 0x846ca68bU;
    return value ^ (value >> 16);
}

} // namespace

std::vector<int32_t> MakeRouting(const std::string &pattern, int32_t rankSize,
    int64_t s, int64_t k, int64_t expertCount, uint32_t seed)
{
    const int64_t n = s * k;
    const int64_t expertsPerRank = expertCount / rankSize;
    std::vector<int32_t> allTopk(static_cast<size_t>(rankSize * n));
    for (int32_t rank = 0; rank < rankSize; ++rank) {
        for (int64_t token = 0; token < s; ++token) {
            int32_t duplicateBase = static_cast<int32_t>(
                Hash(seed + static_cast<uint32_t>(rank * s + token)) % expertCount);
            for (int64_t topk = 0; topk < k; ++topk) {
                const int64_t route = token * k + topk;
                int32_t expert = 0;
                if (pattern == "all_local") {
                    expert = static_cast<int32_t>(rank * expertsPerRank + route % expertsPerRank);
                } else if (pattern == "all_remote") {
                    const int32_t owner = (rank + 1) % rankSize;
                    expert = static_cast<int32_t>(owner * expertsPerRank + route % expertsPerRank);
                } else if (pattern == "duplicate") {
                    expert = topk < k / 2 ? duplicateBase : static_cast<int32_t>(
                        Hash(seed + static_cast<uint32_t>(rank * n + route)) % expertCount);
                } else if (pattern == "biased") {
                    if ((route + rank) % 5 != 0) {
                        expert = static_cast<int32_t>(route % expertsPerRank);
                    } else {
                        expert = static_cast<int32_t>(
                            Hash(seed + static_cast<uint32_t>(rank * n + route)) % expertCount);
                    }
                } else {
                    expert = static_cast<int32_t>((rank * n + route) % expertCount);
                }
                allTopk[static_cast<size_t>(rank * n + route)] = expert;
            }
        }
    }
    return allTopk;
}

bool BuildReferencePlan(int32_t rank, int32_t rankSize, int64_t s, int64_t k,
    int64_t expertCount, const std::vector<int32_t> &allTopk,
    ReferencePlan *plan, std::string *error)
{
    if (plan == nullptr || error == nullptr || rank < 0 || rank >= rankSize ||
        rankSize <= 0 || s <= 0 || k <= 0 || expertCount <= 0 ||
        expertCount % rankSize != 0 ||
        s > std::numeric_limits<int64_t>::max() / k) {
        return false;
    }
    error->clear();
    const int64_t n = s * k;
    const int64_t b = expertCount / rankSize;
    const int64_t nvs = n;
    const uint64_t encodedCapacity =
        static_cast<uint64_t>(rankSize) * static_cast<uint64_t>(nvs);
    if (nvs > std::numeric_limits<int32_t>::max() ||
        encodedCapacity > static_cast<uint64_t>(std::numeric_limits<int32_t>::max()) + 1U) {
        *error = "encoded destination range exceeds int32";
        return false;
    }
    if (static_cast<int64_t>(allTopk.size()) != rankSize * n) {
        *error = "allTopk size mismatch";
        return false;
    }

    std::vector<int32_t> tpe(static_cast<size_t>(rankSize * expertCount), 0);
    for (int32_t source = 0; source < rankSize; ++source) {
        for (int64_t route = 0; route < n; ++route) {
            const int32_t expert = allTopk[static_cast<size_t>(source * n + route)];
            if (expert < 0 || expert >= expertCount) {
                *error = "expert id out of range";
                return false;
            }
            ++tpe[static_cast<size_t>(source * expertCount + expert)];
        }
    }

    std::vector<int32_t> tpePrefix = tpe;
    std::vector<int32_t> expertCountGlobal(static_cast<size_t>(expertCount), 0);
    for (int64_t expert = 0; expert < expertCount; ++expert) {
        int32_t cumulative = 0;
        for (int32_t source = 0; source < rankSize; ++source) {
            cumulative += tpe[static_cast<size_t>(source * expertCount + expert)];
            tpePrefix[static_cast<size_t>(source * expertCount + expert)] = cumulative;
        }
        expertCountGlobal[static_cast<size_t>(expert)] = cumulative;
    }

    std::vector<int32_t> balance(static_cast<size_t>(rankSize), 0);
    for (int32_t home = 0; home < rankSize; ++home) {
        int32_t total = 0;
        for (int64_t localExpert = 0; localExpert < b; ++localExpert) {
            total += expertCountGlobal[static_cast<size_t>(home * b + localExpert)];
        }
        balance[static_cast<size_t>(home)] = total - static_cast<int32_t>(n);
    }

    std::vector<int32_t> z(static_cast<size_t>(rankSize * rankSize), 0);
    while (true) {
        int32_t surplusRank = 0;
        int32_t deficitRank = 0;
        for (int32_t r = 1; r < rankSize; ++r) {
            if (balance[static_cast<size_t>(r)] > balance[static_cast<size_t>(surplusRank)]) {
                surplusRank = r;
            }
            if (balance[static_cast<size_t>(r)] < balance[static_cast<size_t>(deficitRank)]) {
                deficitRank = r;
            }
        }
        if (balance[static_cast<size_t>(surplusRank)] <= 0 ||
            balance[static_cast<size_t>(deficitRank)] >= 0) {
            break;
        }
        const int32_t move = -balance[static_cast<size_t>(deficitRank)];
        z[static_cast<size_t>(surplusRank * rankSize + deficitRank)] = move;
        balance[static_cast<size_t>(surplusRank)] -= move;
        balance[static_cast<size_t>(deficitRank)] = 0;
    }

    std::vector<int32_t> alloc(static_cast<size_t>(expertCount * rankSize), 0);
    for (int64_t expert = 0; expert < expertCount; ++expert) {
        alloc[static_cast<size_t>(AllocIndex(expert, static_cast<int32_t>(expert / b), rankSize))] =
            expertCountGlobal[static_cast<size_t>(expert)];
    }
    for (int32_t home = 0; home < rankSize; ++home) {
        std::vector<int32_t> remaining(static_cast<size_t>(b));
        std::vector<int32_t> quotas(static_cast<size_t>(rankSize));
        for (int64_t localExpert = 0; localExpert < b; ++localExpert) {
            remaining[static_cast<size_t>(localExpert)] =
                expertCountGlobal[static_cast<size_t>(home * b + localExpert)];
        }
        for (int32_t dest = 0; dest < rankSize; ++dest) {
            quotas[static_cast<size_t>(dest)] = z[static_cast<size_t>(home * rankSize + dest)];
        }
        while (true) {
            const int32_t dest = static_cast<int32_t>(
                std::max_element(quotas.begin(), quotas.end()) - quotas.begin());
            if (quotas[static_cast<size_t>(dest)] <= 0) {
                break;
            }
            const int64_t localExpert = std::max_element(remaining.begin(), remaining.end()) - remaining.begin();
            if (remaining[static_cast<size_t>(localExpert)] <= 0) {
                *error = "allocation exhausted before quota";
                return false;
            }
            const int32_t take = std::min(
                remaining[static_cast<size_t>(localExpert)], quotas[static_cast<size_t>(dest)]);
            const int64_t expert = home * b + localExpert;
            alloc[static_cast<size_t>(AllocIndex(expert, dest, rankSize))] += take;
            alloc[static_cast<size_t>(AllocIndex(expert, home, rankSize))] -= take;
            remaining[static_cast<size_t>(localExpert)] -= take;
            quotas[static_cast<size_t>(dest)] -= take;
        }
    }

    std::vector<int32_t> expertOffsets(static_cast<size_t>(rankSize * expertCount), 0);
    std::vector<int32_t> cuAll(static_cast<size_t>(rankSize * (expertCount + b)), 0);
    std::vector<int32_t> expertsToCopy(static_cast<size_t>(rankSize * b), -1);
    std::vector<int32_t> remoteStatsAll(static_cast<size_t>(rankSize * 2), 0);

    for (int32_t dest = 0; dest < rankSize; ++dest) {
        std::vector<int32_t> remote;
        const int64_t localBegin = dest * b;
        const int64_t localEnd = localBegin + b;
        for (int64_t expert = 0; expert < expertCount; ++expert) {
            if ((expert < localBegin || expert >= localEnd) &&
                alloc[static_cast<size_t>(AllocIndex(expert, dest, rankSize))] > 0) {
                remote.push_back(static_cast<int32_t>(expert));
            }
        }
        std::sort(remote.begin(), remote.end(), [&](int32_t lhs, int32_t rhs) {
            const int32_t lhsCount = alloc[static_cast<size_t>(AllocIndex(lhs, dest, rankSize))];
            const int32_t rhsCount = alloc[static_cast<size_t>(AllocIndex(rhs, dest, rankSize))];
            return lhsCount != rhsCount ? lhsCount > rhsCount : lhs > rhs;
        });
        remoteStatsAll[static_cast<size_t>(dest * 2)] = static_cast<int32_t>(remote.size());
        std::vector<uint8_t> selected(static_cast<size_t>(expertCount), 0);
        for (int64_t slot = 0; slot < b && slot < static_cast<int64_t>(remote.size()); ++slot) {
            const int32_t expert = remote[static_cast<size_t>(slot)];
            expertsToCopy[static_cast<size_t>(dest * b + slot)] = expert;
            selected[static_cast<size_t>(expert)] = 1;
            ++remoteStatsAll[static_cast<size_t>((expert / b) * 2 + 1)];
        }

        int32_t start = 0;
        for (int64_t group = 0; group < expertCount + b; ++group) {
            int32_t count = 0;
            int32_t expert = -1;
            if (group < expertCount) {
                if (selected[static_cast<size_t>(group)] == 0) {
                    expert = static_cast<int32_t>(group);
                    count = alloc[static_cast<size_t>(AllocIndex(group, dest, rankSize))];
                }
            } else {
                expert = expertsToCopy[static_cast<size_t>(dest * b + group - expertCount)];
                if (expert >= 0) {
                    count = alloc[static_cast<size_t>(AllocIndex(expert, dest, rankSize))];
                }
            }
            if (count > 0) {
                expertOffsets[static_cast<size_t>(dest * expertCount + expert)] = start;
            }
            start += count;
            cuAll[static_cast<size_t>(dest * (expertCount + b) + group)] = start;
        }
        if (start != nvs) {
            *error = "compact layout does not fill NvS";
            return false;
        }
    }

    std::vector<int32_t> allocPrefix = alloc;
    for (int64_t expert = 0; expert < expertCount; ++expert) {
        int32_t cumulative = 0;
        for (int32_t dest = 0; dest < rankSize; ++dest) {
            cumulative += alloc[static_cast<size_t>(AllocIndex(expert, dest, rankSize))];
            allocPrefix[static_cast<size_t>(AllocIndex(expert, dest, rankSize))] = cumulative;
        }
    }

    std::vector<int32_t> localCount(static_cast<size_t>(expertCount), 0);
    std::vector<int32_t> dst(static_cast<size_t>(n));
    for (int64_t token = 0; token < s; ++token) {
        uint64_t seenLow = 0;
        uint64_t seenHigh = 0;
        for (int64_t topk = 0; topk < k; ++topk) {
            const int64_t route = token * k + topk;
            const int32_t expert = allTopk[static_cast<size_t>(rank * n + route)];
            const int32_t occurrence = localCount[static_cast<size_t>(expert)]++;
            const int32_t sourcePrefix = rank == 0 ? 0 :
                tpePrefix[static_cast<size_t>((rank - 1) * expertCount + expert)];
            const int32_t globalRank = sourcePrefix + occurrence;
            int32_t dest = 0;
            while (dest < rankSize &&
                allocPrefix[static_cast<size_t>(AllocIndex(expert, dest, rankSize))] <= globalRank) {
                ++dest;
            }
            if (dest >= rankSize) {
                *error = "destination not found";
                return false;
            }
            const int32_t previous = dest == 0 ? 0 :
                allocPrefix[static_cast<size_t>(AllocIndex(expert, dest - 1, rankSize))];
            const int32_t raw = static_cast<int32_t>(dest * nvs +
                expertOffsets[static_cast<size_t>(dest * expertCount + expert)] + globalRank - previous);
            bool duplicate = false;
            if (dest < 64) {
                const uint64_t bit = static_cast<uint64_t>(1) << dest;
                duplicate = (seenLow & bit) != 0;
                seenLow |= bit;
            } else {
                const uint64_t bit = static_cast<uint64_t>(1) << (dest - 64);
                duplicate = (seenHigh & bit) != 0;
                seenHigh |= bit;
            }
            dst[static_cast<size_t>(route)] = duplicate ? -raw - 1 : raw;
        }
    }

    plan->dispatchedCapacity = nvs;
    plan->dst = std::move(dst);
    plan->expertsToCopy = std::move(expertsToCopy);
    plan->cuSeqlens.assign(
        cuAll.begin() + rank * (expertCount + b), cuAll.begin() + (rank + 1) * (expertCount + b));
    plan->remoteStats.assign(
        remoteStatsAll.begin() + rank * 2, remoteStatsAll.begin() + (rank + 1) * 2);
    return true;
}

} // namespace TileXRMoonEpTest

File: tools/moonep/planner_reference.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class ReferencePlan:
    dispatched_capacity: int
    dst: tuple[int, ...]
    cu_seqlens: tuple[int, ...]
    zero_fill_ranges: tuple[int, ...]
    experts_to_copy: tuple[int, ...]
    remote_stats: tuple[int, ...]


def deterministic_all_topk(
    rank_size: int, tokens_per_rank: int, topk: int, expert_count: int
) -> tuple[int, ...]:
    route_count = tokens_per_rank * topk
    return tuple(
        (route + source * topk) % expert_count
        for source in range(rank_size)
        for route in range(route_count)
    )


def _first_max(values: Sequence[int]) -> int:
    return max(range(len(values)), key=lambda index: (values[index], -index))


def _first_min(values: Sequence[int]) -> int:
    return min(range(len(values)), key=lambda index: (values[index], index))


def build_reference_plan(
    *,
    rank: int,
    rank_size: int,
    tokens_per_rank: int,
    topk: int,
    expert_count: int,
    all_topk: Sequence[int],
    prefetch_slots: int | None = None,
    token_padding: int = 1,
) -> ReferencePlan:
    if (
        rank_size <= 0
        or rank < 0
        or rank >= rank_size
        or tokens_per_rank <= 0
        or topk <= 0
        or expert_count <= 0
        or expert_count % rank_size != 0
    ):
        raise ValueError("invalid Planner reference dimensions")
    route_count = tokens_per_rank * topk
    experts_per_rank = expert_count // rank_size
    slots = experts_per_rank if prefetch_slots is None else int(prefetch_slots)
    padding = int(token_padding)
    if slots <= 0 or slots > experts_per_rank:
        raise ValueError("prefetch_slots must be in [1, E/R]")
    if padding <= 0:
        raise ValueError("token_padding must be positive")
    dispatched_capacity = route_count + (padding - 1) * 2 * experts_per_rank
    if len(all_topk) != rank_size * route_count:
        raise ValueError("all_topk size does not match rank_size * S * K")
    if rank_size * dispatched_capacity > 2**31:
        raise OverflowError("encoded destination range exceeds signed int32")

    tpe = [[0 for _ in range(expert_count)] for _ in range(rank_size)]
    for source in range(rank_size):
        for route in range(route_count):
            expert = int(all_topk[source * route_count + route])
            if expert < 0 or expert >= expert_count:
                raise ValueError(f"expert id {expert} is outside [0, {expert_count})")
            tpe[source][expert] += 1

    tpe_prefix = [row[:] for row in tpe]
    global_counts = [0 for _ in range(expert_count)]
    for expert in range(expert_count):
        cumulative = 0
        for source in range(rank_size):
            cumulative += tpe[source][expert]
            tpe_prefix[source][expert] = cumulative
        global_counts[expert] = cumulative

    balance = []
    for home in range(rank_size):
        begin = home * experts_per_rank
        balance.append(
            sum(global_counts[begin : begin + experts_per_rank]) - route_count
        )
    moves = [[0 for _ in range(rank_size)] for _ in range(rank_size)]
    while True:
        surplus = _first_max(balance)
        deficit = _first_min(balance)
        if balance[surplus] <= 0 or balance[deficit] >= 0:
            break
        count = -balance[deficit]
        moves[surplus][deficit] = count
        balance[surplus] -= count
        balance[deficit] = 0

    allocation = [[0 for _ in range(rank_size)] for _ in range(expert_count)]
    for expert, count in enumerate(global_counts):
        allocation[expert][expert // experts_per_rank] = count
    for home in range(rank_size):
        begin = home * experts_per_rank
        remaining = global_counts[begin : begin + experts_per_rank]
        quotas = moves[home][:]
        while True:
            destination = _first_max(quotas)
            if quotas[destination] <= 0:
                break
            local_expert = _first_max(remaining)
            if remaining[local_expert] <= 0:
                raise RuntimeError("allocation exhausted before migration quota")
            count = min(remaining[local_expert], quotas[destination])
            expert = begin + local_expert
            allocation[expert][destination] += count
            allocation[expert][home] -= count
            remaining[local_expert] -= count
            quotas[destination] -= count

    expert_offsets = [
        [0 for _ in range(expert_count)] for _ in range(rank_size)
    ]
    all_cu = [
        [0 for _ in range(expert_count + slots)]
        for _ in range(rank_size)
    ]
    all_zero_fill = [
        [[0, 0] for _ in range(expert_count + slots)]
        for _ in range(rank_size)
    ]
    experts_to_copy = [
        [-1 for _ in range(slots)] for _ in range(rank_size)
    ]
    all_remote_stats = [[0, 0] for _ in range(rank_size)]
    for destination in range(rank_size):
        local_begin = destination * experts_per_rank
        local_end = local_begin + experts_per_rank
        remote = [
            expert
            for expert in range(expert_count)
            if not local_begin <= expert < local_end
            and allocation[expert][destination] > 0
        ]
        remote.sort(
            key=lambda expert: (allocation[expert][destination], expert),
            reverse=True,
        )
        all_remote_stats[destination][0] = len(remote)
        selected = [False for _ in range(expert_count)]
        for slot, expert in enumerate(remote[:slots]):
            experts_to_copy[destination][slot] = expert
            selected[expert] = True
            all_remote_stats[expert // experts_per_rank][1] += 1

        end = 0
        for group in range(expert_count + slots):
            count = 0
            expert = -1
            if group < expert_count:
                if not selected[group]:
                    expert = group
                    count = allocation[expert][destination]
            else:
                expert = experts_to_copy[destination][group - expert_count]
                if expert >= 0:
                    count = allocation[expert][destination]
            if count > 0:
                expert_offsets[destination][expert] = end
            padded = ((count + padding - 1) // padding) * padding if count > 0 else 0
            if padded > count:
                all_zero_fill[destination][group] = [end + count, padded - count]
            end += padded
            all_cu[destination][group] = end
        if end > dispatched_capacity:
            raise RuntimeError("padded layout exceeds NvS")

    allocation_prefix = [row[:] for row in allocation]
    for expert in range(expert_count):
        cumulative = 0
        for destination in range(rank_size):
            cumulative += allocation[expert][destination]
            allocation_prefix[expert][destination] = cumulative

    local_counts = [0 for _ in range(expert_count)]
    dst = []
    for token in range(tokens_per_rank):
        seen_destinations: set[int] = set()
        for topk_index in range(topk):
            route = token * topk + topk_index
            expert = int(all_topk[rank * route_count + route])
            occurrence = local_counts[expert]
            local_counts[expert] += 1
            source_prefix = 0 if rank == 0 else tpe_prefix[rank - 1][expert]
            global_occurrence = source_prefix + occurrence
            destination = 0
            while (
                destination < rank_size
                and allocation_prefix[expert][destination] <= global_occurrence
            ):
                destination += 1
            if destination >= rank_size:
                raise RuntimeError("destination not found")
            previous = (
                0
                if destination == 0
                else allocation_prefix[expert][destination - 1]
            )
            raw = (
                destination * dispatched_capacity
                + expert_offsets[destination][expert]
                + global_occurrence
                - previous
            )
            duplicate = destination in seen_destinations
            seen_destinations.add(destination)
            dst.append(-raw - 1 if duplicate else raw)

    return ReferencePlan(
        dispatched_capacity=dispatched_capacity,
        dst=tuple(dst),
        cu_seqlens=tuple(all_cu[rank]),
        zero_fill_ranges=tuple(
            value for group in all_zero_fill[rank] for value in group
        ),
        experts_to_copy=tuple(
            expert for row in experts_to_copy for expert in row
        ),
        remote_stats=tuple(all_remote_stats[rank]),
    )
